From 4817c8b37f7daabaa97a08a0128a64a74eac82f8 Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@freescale.com>
Date: Thu, 20 Oct 2011 19:15:29 +0300
Subject: [PATCH 007/128] dpaa_eth: Don't recycle unaligned buffers

According to the DPAA reference manual, on Rx FMan should only use
16-byte aligned buffers. If we recycle skbs received from the stack,
it's possible to add an unaligned buffer in the BMan pool.
Add an extra check for alignment when determining whether a buffer
is recycleable.

Signed-off-by: Ioana Radulescu <ruxandra.radulescu@freescale.com>
[Kevin: Original patch taken from FSL
QorIQ-SDK-V1.2-SOURCE-20120614-yocto.iso image.]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/freescale/dpa/dpaa_eth.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/dpa/dpaa_eth.c b/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
index 3690db8..73e220c 100644
--- a/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
+++ b/drivers/net/ethernet/freescale/dpa/dpaa_eth.c
@@ -1319,8 +1319,17 @@ static int __hot dpa_tx(struct sk_buff *skb, struct net_device *net_dev)
 	fd.length20 = skb->len;
 	fd.offset = DPA_BP_HEAD; /* This is now guaranteed */
 
+	/*
+	 * Check if skb can be recycled / buffer can be readded to the bman pool
+	 * The following conditions must be met:
+	 * - skb not cloned, not shared
+	 * - buffer size is large enough to accomodate a MTU-sized frame
+	 * - there's enough room in the buffer pool
+	 * - address of the recycled buffer is 16 byte aligned (as per DPAA RM)
+	 */
 	if (likely(skb_is_recycleable(skb, dpa_bp->size + pad)
-			&& (*percpu_priv->dpa_bp_count + 1 <= dpa_bp->count))) {
+			&& (*percpu_priv->dpa_bp_count + 1 <= dpa_bp->count)
+			&& (IS_ALIGNED((unsigned long)skbh, 16)))) {
 		fd.cmd |= FM_FD_CMD_FCO;
 		fd.bpid = dpa_bp->bpid;
 		skb_recycle(skb);
-- 
1.7.9.7

